art/attacks/extraction/copycat_cnn.py

```diff
@@ -0,0 +1,151 @@
+        :type nb_stolen: `int`
+        """
+        super(CopycatCNN, self).__init__(classifier=classifier)
+
+        params = {'batch_size': batch_size,
+                  'nb_epochs': nb_epochs,
+                  'nb_stolen': nb_stolen}
+        self.set_params(**params)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate a thieved classifier.
+
+        :param x: An array with the source input to the victim classifier.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,). Not used in this attack.
+        :type y: `np.ndarray` or `None`
+        :param thieved_classifier: A thieved classifier to be stolen.
+        :type thieved_classifier: :class:`.Classifier`
+        :return: The stolen classifier.
+        :rtype: :class:`.Classifier`
+        """
+        # Warning to users if y is not None
+        if y is not None:
+            logger.warning("This attack does not use the provided label y.")
+
+        # Check the size of the source input vs nb_stolen
+        if x.shape[0] < self.nb_stolen:
+            logger.warning("The size of the source input is smaller than the number of expected stolen examples.")
+
+        # Check if there is a thieved classifier provided for training
+        thieved_classifier = kwargs.get('thieved_classifier')
+        if thieved_classifier is None or not isinstance(thieved_classifier, Classifier):
+            raise ValueError('A thieved classifier is needed.')
+
+        # Select data to attack
+        selected_x = self._select_data(x)
+
+        # Query the victim classifier
+        fake_labels = self._query_label(selected_x)
+
+        # Train the thieved classifier
+        thieved_classifier.fit(x=selected_x, y=fake_labels, batch_size=self.batch_size, nb_epochs=self.nb_epochs)
+
+        return thieved_classifier
+
+    def _select_data(self, x):
+        """
+        Select data to attack.
+
+        :param x: An array with the source input to the victim classifier.
+        :type x: `np.ndarray`
+        :return: An array with the selected input to the victim classifier.
+        :rtype: `np.ndarray`
+        """
+        nb_stolen = np.minimum(self.nb_stolen, x.shape[0])
+        rnd_index = np.random.choice(x.shape[0], nb_stolen, replace=False)
+
+        return x[rnd_index].astype(NUMPY_DTYPE)
+
+    def _query_label(self, x):
+        """
+        Query the victim classifier.
+
+        :param x: An array with the source input to the victim classifier.
+        :type x: `np.ndarray`
+        :return: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes).
+        :rtype: `np.ndarray`
+        """
+        labels = self.classifier.predict(x=x, batch_size=self.batch_size)
+        labels = np.argmax(labels, axis=1)
+        labels = to_categorical(labels=labels, nb_classes=self.classifier.nb_classes())
```

Comment:
Is this specific to the paper cited? I would rather see the option to return soft labels which makes some of these attacks more effective. Perhaps as an argument to the `__init__` or `generate`. The ability to handle the soft labels depends on the `thieved_classifier`.

Reply:
Thanks @imolloy , this is also a cool idea to keep a soft version of labels. However, as discussed with @beat-buesser , the thieved_classifier which is expected to be an ART classifier and therefore only accepts 1-hot labels, we temporarily keep this type of hard labels. We need to investigate more if changing ART classifier's fit function will cause any trouble for ART. See issue #251